File: codility/26.MaxSliceSum.java
public class MaxSliceSum {
    
}
// 最大子串和 O(n)
// Codility
class Solution {
    public int solution(int[] A) {
        // write your code in Java SE 8

        // 不允许空 为子集
        int maxSum = A[0];
        int currentSum = A[0];
        for(int i = 1;i<A.length;i++){
            // 如果允许空串
            // if(currentSum < 0){
            //     currentSum = 0;
            // } 
            // currentSum += A[i];
            // 如果不允许空串，负值也要计算其中
            if(currentSum <0){
                currentSum = A[i];
            } else {
                currentSum += A[i];
            }
                
            if(currentSum > maxSum ){
                maxSum = currentSum;
            }
        }
        return maxSum; 
    }
}

File: codility/26.MaxSliceSum.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(A):
    # Implement your solution here
    # Same as Leetcode 53. Maximum Subarray
    dq = [0] * len(A)
    dq[0] = A[0]
    # max = A[0]
    for i in range(1, len(A)):
        if dq[i -1] <= 0:
            dq[i] = A[i]
        else:
            dq[i] = dq[i-1] + A[i]
    return max(dq)